src/main/java/org/opensearch/knn/index/codec/KNN80Codec/KNN80DocValuesConsumer.java

```diff
@@ -110,61 +111,72 @@ public void addKNNBinaryField(FieldInfo field, DocValuesProducer valuesProducer,
-        String engineFileName;
-        String indexPath;
-        NativeIndexCreator indexCreator;
-        final KNNEngine knnEngine = getKNNEngine(field);
-        if (field.attributes().containsKey(MODEL_ID)) {
+        KNNCodecUtil.Pair pair = null;
+        try {
+            pair = KNNCodecUtil.getFloats(values);
+            if (pair.getVectorAddress() == 0 || pair.docs.length == 0) {
+                logger.info("Skipping engine index creation as there are no vectors or docs in the segment");
+                return;
+            }
+            long arraySize = calculateArraySize(pair.docs.length, pair.getDimension(), pair.serializationMode);
+            if (isMerge) {
+                KNNGraphValue.MERGE_CURRENT_OPERATIONS.increment();
+                KNNGraphValue.MERGE_CURRENT_DOCS.incrementBy(pair.docs.length);
+                KNNGraphValue.MERGE_CURRENT_SIZE_IN_BYTES.incrementBy(arraySize);
+            }
+            // Increment counter for number of graph index requests
+            KNNCounter.GRAPH_INDEX_REQUESTS.increment();
+            // Create library index either from model or from scratch
+            String engineFileName;
+            String indexPath;
+            NativeIndexCreator indexCreator;
+            final KNNEngine knnEngine = getKNNEngine(field);
+            if (field.attributes().containsKey(MODEL_ID)) {
+
+                String modelId = field.attributes().get(MODEL_ID);
+                Model model = ModelCache.getInstance().get(modelId);
+
+                engineFileName = buildEngineFileName(state.segmentInfo.name, knnEngine.getVersion(), field.name, knnEngine.getExtension());
+                indexPath = Paths.get(((FSDirectory) (FilterDirectory.unwrap(state.directory))).getDirectory().toString(), engineFileName)
+                    .toString();
+
+                if (model.getModelBlob() == null) {
+                    throw new RuntimeException("There is no trained model with id \"" + modelId + "\"");
+                }
 
-            String modelId = field.attributes().get(MODEL_ID);
-            Model model = ModelCache.getInstance().get(modelId);
+                KNNCodecUtil.Pair finalPair = pair;
+                indexCreator = () -> createKNNIndexFromTemplate(model.getModelBlob(), finalPair, knnEngine, indexPath);
+            } else {
 
-            engineFileName = buildEngineFileName(state.segmentInfo.name, knnEngine.getVersion(), field.name, knnEngine.getExtension());
-            indexPath = Paths.get(((FSDirectory) (FilterDirectory.unwrap(state.directory))).getDirectory().toString(), engineFileName)
-                .toString();
+                engineFileName = buildEngineFileName(state.segmentInfo.name, knnEngine.getVersion(), field.name, knnEngine.getExtension());
+                indexPath = Paths.get(((FSDirectory) (FilterDirectory.unwrap(state.directory))).getDirectory().toString(), engineFileName)
+                    .toString();
```

Comment:
can these two lines be extracted in a method? I think it's a duplication of the same code you've added to the same class on lines 139-140